xds/src/main/java/io/grpc/xds/GcpAuthenticationFilter.java

```diff
@@ -0,0 +1,146 @@
+
+  }
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfig(Message rawProtoMessage) {
+    GcpAuthnFilterConfig gcpAuthnProto;
+    if (!(rawProtoMessage instanceof Any)) {
+      return ConfigOrError.fromError("Invalid config type: " + rawProtoMessage.getClass());
+    }
+    Any anyMessage = (Any) rawProtoMessage;
+
+    try {
+      gcpAuthnProto = anyMessage.unpack(GcpAuthnFilterConfig.class);
+    } catch (InvalidProtocolBufferException e) {
+      return ConfigOrError.fromError("Invalid proto: " + e);
+    }
+
+    // Validate cache_config
+    TokenCacheConfig cacheConfig = gcpAuthnProto.getCacheConfig();
+    if (cacheConfig != null) {
+      long cacheSize = cacheConfig.getCacheSize().getValue();
+      if (cacheSize <= 0) {
+        return ConfigOrError.fromError(
+            "cache_config.cache_size must be in the range (0, INT64_MAX)");
+      }
+    }
+
+    // Create and return the GcpAuthenticationConfig
+    GcpAuthenticationConfig config = new GcpAuthenticationConfig(
+        gcpAuthnProto.getHttpUri().getUri(),
+        cacheConfig != null ? cacheConfig.getCacheSize().getValue() : 10
+    );
+    return ConfigOrError.fromConfig(config);
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfigOverride(Message rawProtoMessage) {
+    return parseFilterConfig(rawProtoMessage);
+  }
+
+  @Nullable
+  @Override
+  public ClientInterceptor buildClientInterceptor(FilterConfig config,
+      @Nullable FilterConfig overrideConfig, PickSubchannelArgs args,
+      ScheduledExecutorService scheduler) {
+
+    ComputeEngineCredentials credentials = ComputeEngineCredentials.create();
+    LruCache<String, CallCredentials> callCredentialsCache =
+        new LruCache<>(((GcpAuthenticationConfig) config).getCacheSize());
+    return new ClientInterceptor() {
+      @Override
+      public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+          MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+
+        /*String clusterName = callOptions.getOption(InternalXdsAttributes.ATTR_CLUSTER_NAME);
+        if (clusterName == null) {
+          return next.newCall(method, callOptions);
+        }*/
+
+        // TODO: Fetch the CDS resource for the cluster.
+        // If the CDS resource is not available, fail the RPC with Status.UNAVAILABLE.
+
+        // TODO: Extract the audience from the CDS resource metadata.
+        // If the audience is not found or is in the wrong format, fail the RPC.
+        String audience = "TEST_AUDIENCE";
+
+        try {
+          CallCredentials callCredentials =
+              getCallCredentials(callCredentialsCache, audience, credentials);
+          callOptions = callOptions.withCallCredentials(callCredentials);
+        } catch (Exception e) {
+          throw new RuntimeException("Failed to attach CallCredentials.", e);
```

Comment:
Interceptors should not purposefully throw. Instead, you can return a FailingClientCall or add a CallCredentials that fails the RPC. But I actually don't see how any of this will fail, so don't know exactly what is being handled.